cdap-ui/app/directives/my-pipeline-configurations/my-batch-pipeline-config/my-batch-pipeline-config-ctrl.js

```diff
@@ -0,0 +1,125 @@
+* the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations under
+* the License.
+*/
+
+class MyBatchPipelineConfigCtrl {
+  constructor(uuid, HydratorPlusPlusHydratorService, HYDRATOR_DEFAULT_VALUES) {
+    this.uuid = uuid;
+    this.HydratorPlusPlusHydratorService = HydratorPlusPlusHydratorService;
+
+    this.engine = this.store.getEngine();
+    this.engineForDisplay = this.engine === 'mapreduce' ? 'MapReduce' : 'Spark';
+    this.instrumentation = this.store.getInstrumentation();
+    this.stageLogging = this.store.getStageLogging();
+    this.numRecordsPreview = this.store.getNumRecordsPreview();
+    this.numberConfig = {
+      'widget-attributes': {
+        min: 0,
+        default: HYDRATOR_DEFAULT_VALUES.numOfRecordsPreview,
+        showErrorMessage: false,
+        convertToInteger: true
+      }
+    };
+
+    this.customEngineConfig = {
+      'pairs': HydratorPlusPlusHydratorService.convertMapToKeyValuePairs(this.store.getCustomConfig())
+    };
+
+    this.activeTab = 'runtimeArgs';
+    this.runtimeArguments = this.checkForReset(this.runtimeArguments);
+    this.onRuntimeArgumentsChange = this.onRuntimeArgumentsChange.bind(this);
+    this.onCustomEngineConfigChange = this.onCustomEngineConfigChange.bind(this);
+    this.getResettedRuntimeArgument = this.getResettedRuntimeArgument.bind(this);
+  }
+
+  onRuntimeArgumentsChange(newRuntimeArguments) {
+    this.runtimeArguments = this.checkForReset(newRuntimeArguments);
+  }
+
+  onCustomEngineConfigChange(newCustomConfig) {
+    this.customEngineConfig = newCustomConfig;
+  }
+
+  onEngineChange() {
+    this.engineForDisplay = this.engine === 'mapreduce' ? 'MapReduce' : 'Spark';
+  }
+
+  addCustomConfig() {
+    this.customEngineConfig.pairs.push({
+      key: '',
+      value: '',
+      uniqueId: 'id-' + this.uuid.v4()
+    });
+  }
+
+  checkForReset(runtimeArguments) {
+    let runtimeArgumentsPairs = runtimeArguments.pairs;
+    this.resettingFields = false;
+    for (let i = 0; i < runtimeArgumentsPairs.length; i++) {
+      if (runtimeArgumentsPairs[i].notDeletable) {
+        let runtimeArgKey = runtimeArgumentsPairs[i].key;
+        if (this.resolvedMacros.hasOwnProperty(runtimeArgKey)) {
+          if (this.resolvedMacros[runtimeArgKey] !== runtimeArgumentsPairs[i].value) {
+            runtimeArgumentsPairs[i].showReset = true;
+          } else {
+            runtimeArgumentsPairs[i].showReset = false;
+          }
+        }
+      }
+    }
+    return runtimeArguments;
+  }
+
+  getResettedRuntimeArgument(index) {
+    let runtimeArgKey = this.runtimeArguments.pairs[index].key;
+    this.resettingFields = true;
```

Comment:
The idea of using store is to communicate between components. Right now what we are doing is having a flag which we update to invoke `componentWillReceiveProps` function which updates the `KeyValueStore` which inturn re-renders the component. Ideally it should have been the other way. We update the store and the component automatically updates.